Add automatic cluster upgrades RFC (#15)
giantswarm/rfc: automatic-cluster-upgrades/README.md

# RFC 0015 - Automatic workload cluster upgrades

## User stories

> - As a Giant Swarm customer, and also as Giant Swarm, we would like to have workload clusters running in the latest version automatically, and not worry about patching security vulnerabilities or cluster components running out of updates (e. g. Kubernetes release end of life).

Author: ```suggestion
- As a Giant Swarm customer, and also as a Giant Swarm maintainer, we would like to have workload clusters running in the latest version automatically, and not worry about patching security vulnerabilities or cluster components running out of updates (e. g. Kubernetes release end of life).
```

> - As a customer admin, I would like to define maintenance windows to ensure workload clusters are only upgraded within these hours.

Author: ```suggestion
- As a customer admin, I would like to define upgrade windows, to ensure that upgrades happen when disruption is most permittable and impact on production traffic is minimal.
```

- As a customer, and also as Giant Swarm, we want to capture the result of an upgrade. Based on the result we can alert on a certain condition, track the history for certain cluster(s) or automate response based on the end state.

## Reasoning

Our customers need to plan every upgrade on their side and manage the upgrades for every workload cluster release. Each customer has their own requirements regarding when they need to schedule upgrades (freeze periods, maintenance windows, etc.). At the same time, cloud native projects and Giant Swarm are continuously providing new releases. The preparation and execution of upgrades is time and labor intensive, while the fact that they are executed manually does not add much value to stakeholders.

Giant Swarm designed a versioning scheme (based on Semver), which includes all components on the stack, to ensure idempotence and expose the level of impact that changes bring to the customer. Patches and Minor releases should have no impact on the customer workloads, which means that they can be applied in the environments at any time.

## Current state

> Right now the automatic upgrades are carried by Solution Engineers agreeing with the customer on upgrade windows and environments selection. The idea is to automate all this behavior and information under a Kubernetes operator(s) and Custom Resource(s).

Author: ```suggestion
Right now the automatic upgrades are carried by Solution Engineers agreeing with the customer on upgrade windows and environments selected. The idea is to automate all this behaviour and information under a Kubernetes operator(s) and Custom Resource(s).
```

> The Solution engineer plans cluster upgrades based on cluster type (environment, team,...) and maintenance windows offered by the customers. The upgrade is just a change on the cluster version field on the cluster Custom Resource (CR).

Author: ```suggestion
The Solution Engineer plans cluster upgrades based on cluster type (environment, team,...) and maintenance windows offered by the customers. The upgrade is just a change on the cluster version field on the cluster Custom Resource (CR).
```

#### Cluster upgrade scheduler

Ideally to automate the whole process there will be an operator that creates all the `Cluster Upgrade` CRs when a new release of Giant Swarm is created (and/or possibly other events).

#### Existing CAPI controllers

There was [a proposal in CAPI upstream](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191017-kubeadm-based-control-plane.md) for the KubeadmControlPlane (later included in CAPZ components) to add `upgradeAfter` parameter to influence the upgrades. Though this field only means a cluster upgrade will be triggered after that date, to force an upgrade, it is related to the goal of controlling upgrades. The idea would be to create a new proposal that allows pointing a new entity (`MaintenanceSchedule `/`UpgradePolicy`)  to extend the possibilities more than a single timestamp. Later controllers could leverage on that to allow changes or not to their reconciled resources.

## Open questions

> - Do we scheduled automatic upgrades by default? or we let customer to scheduled them (maybe suggesting in our UI a cluster upgrade need to be scheduled)? 

Reviewer: ```suggestion
- Do we schedule automatic upgrades by default? Or do we let customer schedule them (maybe suggesting in our UI a cluster upgrade need to be scheduled)?
```

> - Do we setup a default maintenance schedule for cluster that does not have one defined/attached?

Reviewer: ```suggestion
- Do we set up a default maintenance schedule for a cluster that does not have one defined/attached?
```

> - If we let customer to freeze the cluster we can apply for all type of upgrades? at some point it can be used for chart operator (or a new app upgrade operator) too to disable app upgrades?

Reviewer: ```suggestion
- If we allow a customer to freeze the cluster, can apply all types of upgrades? Can this mechanism also be used at some point by chart-operator (or a new app upgrade operator) to disable app upgrades?
```

## Additional actions

> - Adapt our monitoring to not trigger false positive during upgrades. Right now, the upgrades usually trigger alerts (like `ServiceLevelBurnRateTooHigh`) which are false positives (all nodes are rolled during upgrade, so it should be omitted). At the same time having specific alerts for upgrades can be valuable (like one when master is not coming up after X minutes)

Reviewer: ```suggestion
- Adapt our monitoring to not trigger false positives during upgrades. Right now, the upgrades usually trigger alerts (like `ServiceLevelBurnRateTooHigh`) which are false positives (all nodes are rolled during upgrade, so it should be omitted). At the same time having specific alerts for upgrades can be valuable (like one when master is not coming up after X minutes)
```
